### PR TITLE
Add an AbstractStorageBasedTypeProvider

### PR DESCRIPTION
**Pull request by J-N-K**
Related to #3394 

This adds a combined type provider for use in bindings that create `ChannelType`s, `ThingType`s or `ChannelGroupType`s dynamically. It provides methods for adding/removing these types and persists them, so they are available on next start-up. Together with #3397 this should allow a smooth transition for all bindings.

There are two reference implementations for use in bindings: https://github.com/openhab/openhab-addons/pull/14501 and https://github.com/openhab/openhab-addons/pull/14507

### CODE REVIEW

**Comment by lolodomo** [comment]
You are pointing twice the same PR. Please fix one of them to point 14701.

**Comment by splatch** [comment]
One note about naming - its not only a dynamic provider, its storage based. The dynamism part is far less specific than the way how it retains its sate.

**Comment by digitaldan** [comment]
This is a list of openhab-addons that use ChannelTypeBuilder, i am guesing these are the likely addons that need updating?

- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/Tr064ChannelTypeProvider.java#L25
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java#L49
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.networkupstools/src/main/java/org/openhab/binding/networkupstools/internal/NUTDynamicChannelFactory.java#L25
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.guntamatic/src/main/java/org/openhab/binding/guntamatic/internal/GuntamaticChannelTypeProvider.java#L23
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/ChannelsTypeProviderPreset.java#L31
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/providers/NetatmoChannelTypeProvider.java#L28
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java#L50
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/BasicChannelTypeProvider.java#L29
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/internal/SmartMeterChannelTypeProvider.java#L29
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/handler/KM200ThingHandler.java#L51
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java#L46
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.solarwatt/src/main/java/org/openhab/binding/solarwatt/internal/channel/SolarwattChannelTypeProvider.java#L29
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/providers/DsChannelTypeProvider.java#L37
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/ComponentChannel.java#L39
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/providers/ResolChannelTypeProvider.java#L24
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/handler/HarmonyDeviceHandler.java#L40
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.remoteopenhab/src/main/java/org/openhab/binding/remoteopenhab/internal/handler/RemoteopenhabBridgeHandler.java#L74
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.wundergroundupdatereceiver/src/main/java/org/openhab/binding/wundergroundupdatereceiver/internal/WundergroundUpdateReceiverUnknownChannelTypeProvider.java#L29
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/CharacteristicChannelTypeProvider.java#L33
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/ChannelsTypeProviderAvailableInputs.java#L31
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/itests/org.openhab.binding.ntp.tests/src/main/java/org/openhab/binding/ntp/test/NtpOSGiTest.java#L68
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/itests/org.openhab.binding.wemo.tests/src/main/java/org/openhab/binding/wemo/internal/test/GenericWemoOSGiTest.java#L59
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.sensibo/src/main/java/org/openhab/binding/sensibo/internal/handler/SensiboSkyHandler.java#L61
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.wundergroundupdatereceiver/src/test/java/org/openhab/binding/wundergroundupdatereceiver/internal/WundergroundUpdateReceiverServletTest.java#L75
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/handler/HarmonyHubHandler.java#L49
- https://github.com/openhab/openhab-addons/blob/2ec097d5295364370013f83a6bd82611e2486ac8/bundles/org.openhab.binding.wundergroundupdatereceiver/src/test/java/org/openhab/binding/wundergroundupdatereceiver/internal/WundergroundUpdateReceiverDiscoveryServiceTest.java#L51

**Comment by J-N-K** [comment]
TR-064 is not affected, it dynamically creates the types, but does so when the binding starts up. Only bindings that create the thing/channel-type when the thing initializes are affected. But there is no need to rush, with #3397 in place it only causes a delay on startup but does not prevent the thing from initializing.

**Comment by digitaldan** [comment]
>But there is no need to rush, with https://github.com/openhab/openhab-core/pull/3397 in place it only causes a delay on startup but does not prevent the thing from initializing.

Thanks for the workaround, and fixes!

**Comment by ccutrer** [comment]
Is this PR waiting for something? I'm not sure when I'll have time (my discretionary coding time is much less than it used to be), but I'd like to get started on using this for mqtt.homie and mqtt.homeassistant at some point. I suppose this PR doesn't _have_ to be merged before I do that, it would just be a little bit easier.

**Comment by J-N-K** [comment]
It's waiting for review. From the code it's complete.

**Comment by ccutrer** [comment]
Happy to look, though I'm not a core maintainer. I looked through this PR, as well as the two linked reference implementations.

**Comment by lolodomo** [comment]
I can see that the `locale` parameter of the "get" methods are ignored in your implementation.
Was it handled in the existing abstract thing type / channel type providers ?

**Comment by lolodomo** [comment]
> I can see that the `locale` parameter of the "get" methods are ignored in your implementation.
> Was it handled in the existing abstract thing type / channel type providers ?

It is at least handled in class `DefaultSystemChannelTypeProvider`.

**Comment by lolodomo** [comment]
> > I can see that the `locale` parameter of the "get" methods are ignored in your implementation.
> > Was it handled in the existing abstract thing type / channel type providers ?
> 
> It is at least handled in class `DefaultSystemChannelTypeProvider`.

Maybe I am wrong and for dynamic types, it is rather the responsibility of each binding to add/provide already localized types.

**Comment by J-N-K** [comment]
If we look at channel labels and description for managed things: they are localized during creation and then the label/description is stored/restored from the database. The same thing applies here: The localization should take place during creation of the channel-type/thing-type and is then stored.

**Comment by lolodomo** [comment]
> The localization should take place during creation of the channel-type/thing-type and is then stored.

I think you're right. So `locale` can be ignored, no problem.

File: itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/binding/AbstractStorageBasedTypeProviderOSGiTest.java
Comment (kaikreuzer) [inline itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/binding/AbstractStorageBasedTypeProviderOSGiTest.java:43]:
```suggestion
 * and {@link ThingType}s
```

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks.

**Comment by lolodomo** [comment]
@J-N-K : as mentioned in openhab/openhab-addons@14954, I encounter deserialization errors when using this new class:
```
00:04:16.548 [ERROR] [ore.storage.json.internal.JsonStorage] - Couldn't deserialize value 'org.openhab.core.storage.json.internal.StorageEntry@35e456f3'. Root cause is: Interfaces can't be instantiated! Register an InstanceCreator or a TypeAdapter for this type. Interface name: org.openhab.core.types.StateDescriptionFragment
```

**Comment by J-N-K** [comment]
Do you think you can provide a PR for that?

**Pull request merged**
